lib/Twig/Extensions/Node/Trans.php

```diff
@@ -149,7 +175,136 @@ protected function compileString(Twig_Node $body)
             $msg = $body->getAttribute('data');
         }
 
-        return array(new Twig_Node(array(new Twig_Node_Expression_Constant(trim($msg), $body->getTemplateLine()))), $vars);
+        return array(new Twig_Node(array(new Twig_Node_Expression_Constant($this->normalize ? $this->normalizeString($msg) : trim($msg), $body->getTemplateLine()))), $vars);
+    }
+
+    /**
+     * Normalizes a string (removes spaces inside the string).
+     *
+     * Why?: For large translatable strings or strings spanning multiple lines,
+     * it is necessary to normalize the string so that translators doesn't get
+     * carriage returns or tab characters inside the string. Normalization is
+     * also needed to mantain the integrity of multiline strings and the
+     * indentation of the source file changed. Take for example:
+     *
+     *  Before (A):
+     *
+     *      {% trans %}
+     *          This is a translatable string
+     *          spanning multiple lines
+     *      {% endtrans %}
+     *
+     *  After (B):
+     *
+     *      {% if some_condition %}
+     *          {% trans %}
+     *              This is a translatable string
+     *              spanning multiple lines
+     *          {% endtrans %}
+     *      {% endif %}
+     *
+     * If the above example, we needed to add some condition and thus added
+     * indentation because of the "if". Then, the translatable string changes
+     * and to Gettext it is a different string because of the different spaces
+     * inside the sentence.
+     *
+     * @param string $msg  The message string to be normalized.
+     * @param string $glue The character used to replace spaces.
+     *
+     * @return string The normalized string.
+     */
+    protected function normalizeString($msg, $glue = ' ')
+    {
+        return preg_replace('/\s+/u', $glue, trim($msg));
```

Comment:
this would turn newlines into spaces. That might be totally unexpected

Reply:
Not exactly, it collapses all whitespace characters (those included in the `\s` meta) to a single space. It is not enabled by default anyway, so users that enable the `normalize` feature should already expect this to happen.

The rationale is as follows:

This solves the problem when re-indenting Twig templates, so for example:

    {% if some_condition %}
        {% trans %}
            This is a translatable string
            spanning multiple lines
        {% endtrans %}
    {% endif %}

Gets the correct translatable sentence as `This is a translatable string spanning multiple lines` even if we delete the `{% if %}` and lines get re-indented.

It also solves the problem when we need to change formatting so this:

    {% trans %}String to translate{% endtrans %}

Gets reformatted as:

    {% trans %}
        String to translate
    {% endtrans %}

In both cases we would end up with "fuzzy" translations in the resulting PO files because there was a slight change in spacing. Also this formatting differences can lead to duplicated sentences since the first one would export as `String to translate` but the second one would export as `\n    String to translate\n`.

If your concern is with newlines, consider that 99% of the time sentences will be used in a HTML context in which the browser would end up collapsing spaces anyway.